### ISSUE TITLE
Deis logs not consistent

### ISSUE THREAD

**Issue by anoopknayak** [deis/deis#3051]
I am using Deis 1.3.1 on AWS. I have 3 instances(m3.large) in a cluster. There are about 9 apps deployed on deis and almost all of the apps have multiple process types as Procfile is used for deployment.

My problem is I have a background celery task which works perfectly sometimes and on the other hand it tends to freeze up(I don't get the logs on what's happening there). Also I wanted to know whether I can separate out the logs(as I suspect deis is not be able to log from all the process simultaneous logs from all processes). As a result deis is not stable. A quick snapshot of the logs is as follows:
```
2015-02-10T19:40:09UTC box[ts.1]: [2015-02-10 19:40:09,170: INFO/MainProcess] Received task: Box.v1.tasks.transfer_files_task[f870867a-9d7b-4719-a32d-b7f976519b6a]
2015-02-10T19:40:09UTC box[ts.1]: [2015-02-10 19:40:09,173: INFO/Worker-1] Starting new HTTP connection (1): api-beta.octonius.com
2015-02-10T19:40:09UTC box[ts.1]: [2015-02-10 19:40:09,188: INFO/Worker-1] Starting new HTTP connection (1): api-beta.octonius.com
2015-02-10T19:40:09UTC box[ts.1]: [2015-02-10 19:40:09,199: INFO/Worker-1] Starting new HTTP connection (1): api-beta.octonius.com
2015-02-10T19:40:09UTC box[ts.1]: [2015-02-10 19:40:09,398: INFO/Worker-1] Starting new HTTPS connection (1): api-content.dropbox.com
2015-02-10T19:40:09UTC box[ts.1]: [2015-02-10 19:40:09,805: WARNING/Worker-1] Downloaded from the source cloud
2015-02-10T19:40:09UTC box[ts.1]: [2015-02-10 19:40:09,805: WARNING/Worker-1] Uploading to destination cloud
2015-02-10T19:40:09UTC box[ts.1]: [2015-02-10 19:40:09,806: WARNING/Worker-1] Content Length in MUW: 2243335
2015-02-10T19:40:09UTC box[ts.1]: [2015-02-10 19:40:09,807: INFO/Worker-1] Starting new HTTPS connection (1): upload.box.com
2015-02-10T19:40:10UTC box[web.1]: Webhook data received:
2015-02-10T19:40:10UTC box[web.1]: New file detected
2015-02-10T19:40:10UTC box[web.1]: New file detected and it is in a project
2015-02-10T19:40:10UTC box[web.1]: Going to create new file
2015-02-10T19:40:10UTC box[web.1]: [pid: 21|app: 0|req: 90/102] 10.21.2.168 () {38 vars in 553 bytes} [Tue Feb 10 19:40:10 2015] POST /box/notifications => generated 0 bytes in 470 msecs (HTTP/1.0 200) 2 headers in 78 bytes (1 switches on core 0)
2015-02-10T19:40:16UTC box[web.1]: Webhook data received:
2015-02-10T19:40:16UTC box[web.1]: New file detected
2015-02-10T19:40:16UTC box[web.1]: New file detected and it is in a project
2015-02-10T19:40:16UTC box[web.1]: Going to create new file
2015-02-10T19:40:16UTC box[web.1]: [pid: 21|app: 0|req: 91/103] 10.21.2.168 () {38 vars in 553 bytes} [Tue Feb 10 19:40:16 2015] POST /box/notifications => generated 0 bytes in 510 msecs (HTTP/1.0 200) 2 headers in 78 bytes (1 switches on core 1)
2015-02-10T19:40:17UTC box[web.1]: Webhook data received:
2015-02-10T19:40:17UTC box[web.1]: New file detected
2015-02-10T19:40:17UTC box[web.1]: File not in projects.
2015-02-10T19:40:17UTC box[web.1]: [pid: 21|app: 0|req: 92/104] 10.21.2.168 () {38 vars in 553 bytes} [Tue Feb 10 19:40:17 2015] POST /box/notifications => generated 0 bytes in 7 msecs (HTTP/1.0 200) 2 headers in 78 bytes (1 switches on core 0)
2015-02-10T19:42:48UTC box[web.1]: [pid: 21|app: 0|req: 94/105] 10.21.2.168 () {40 vars in 756 bytes} [Tue Feb 10 19:42:48 2015] POST /box/54da562ef48b82001c82d7a2/get_download_data/ => generated 320 bytes in 450 msecs (HTTP/1.0 200) 2 headers in 72 bytes (1 switches on core 1)
2015-02-10T19:42:49UTC box[web.1]: [pid: 21|app: 0|req: 94/106] 10.21.2.168 () {40 vars in 756 bytes} [Tue Feb 10 19:42:48 2015] POST /box/54da562ef48b82001c82d7a2/get_download_data/ => generated 347 bytes in 506 msecs (HTTP/1.0 200) 2 headers in 72 bytes (1 switches on core 0)
2015-02-10T19:49:02UTC box[beat.1]: [2015-02-10 19:49:02,850: INFO/MainProcess] Scheduler: Sending due task refresh-tokens-every-20-minutes (Box.v1.tasks.refresh_token_service)
2015-02-10T19:49:02UTC box[rts.1]: [2015-02-10 19:49:02,854: INFO/MainProcess] Received task: Box.v1.tasks.refresh_token_service[d2e4b8cc-d1d4-4e6c-95c0-431b30d48af7]
2015-02-10T19:49:02UTC box[rts.1]: [2015-02-10 19:49:02,857: INFO/Worker-1] Starting new HTTP connection (1): api-beta.octonius.com
2015-02-10T19:49:02UTC box[rts.1]: [2015-02-10 19:49:02,906: INFO/MainProcess] Task Box.v1.tasks.refresh_token_service[d2e4b8cc-d1d4-4e6c-95c0-431b30d48af7] succeeded in 0.0503688019817s: 'success! Updated 0 credentials at 2015-02-10 19:49:02.895593'
2015-02-10T20:00:11UTC box[web.1]: Webhook data received:
```

You can see that *ts.1* doesn't output log even after 5 minutes(assuming it failed due to the celery hard limit on tasks). If it was proceeding I should get further logs(which I do when I run the same task with the same parameters sometime else..). I thought it might be due to memory being full, but my development system has way less RAM and I run it as a docker container with no problems for 12 hrs straight.

I tried setting up Platform logging(to paper trail) but it was failing before(i was using deis 1.2.2 then) and so gave up on it before. Is there any way on accessing the logs directly from the container? Or is it because deis is only designed to run on the "web" or "cmd" process alone? I currently have "web" and four other celery based background workers on a single app. And two of them are scaled twice.

**Comment by bacongobbler** [comment]
> as I suspect deis is not be able to log from all the process simultaneous logs from all processes

logspout is the component behind the log aggregation. It bind-mounts the host's docker socket, listens for events and pipes all container logs over to deis-logger, so the platform and application logs should all be there. Since we provision it as a global unit, *every* host's docker container logs are shipped if everything is working correctly.

If you want logs directly from a container, `deis logs` retrieves the application logs from ceph (deis-logger is backed by cephfs). If you're not seeing any logs, then I'd inspect what logspout and logger are up to. If they're down then logs won't get shipped.

**Comment by anoopknayak** [comment]
The output I pasted before is on using `deis logs`... Is there anyway where I can get the logs of the individual containers(or processes in this case *ts.1*)?

**Comment by bacongobbler** [comment]
From my previous comment...

> If you're not seeing any logs, then I'd check and see what logspout and logger are up to. If they're down then logs won't get shipped.

You can still use `fleetctl` to check the health of your apps, though the better solution would be to figure out why the logs are not getting shipped.

**Comment by carmstrong** [comment]
> Is there anyway where I can get the logs of the individual containers(or processes in this case ts.1 )?

Nope - `deis logs` will show you the logs for all processes for an app.

> I tried setting up Platform logging(to paper trail) but it was failing before(i was using deis 1.2.2 then) and so gave up on it before.

You should give this another shot - we have several users who are shipping their logs using this type of method. Previously, long log messages would get dropped due to reaching the maximum size of a UDP packet, but the documentation now uses TLS: http://docs.deis.io/en/latest/managing_deis/platform_logging/

> Or is it because deis is only designed to run on the "web" or "cmd" process alone? I currently have "web" and four other celery based background workers on a single app. And two of them are scaled twice.

This is perfectly fine. Plenty of folks run all sorts of applications with custom Proctypes. Only web and cmd types are picked up by the router, though.

**Comment by anoopknayak** [comment]
Thanks for the prompt responses. I just set up paper trail..All seems to be fine now. And will try with fleetctl to see what is happening with that unit and come back

**Comment by carmstrong** [comment]
> All seems to be fine now. And will try with fleetctl to see what is happening with that unit and come back

Sounds like a plan!

**Comment by carmstrong** [comment]
Also, you can always enter a container (app containers included) with `nse`. For example: `nse myapp_v3.web` -- this will allow you to poke around and see what's going on, which is helpful for debugging if you're not getting logs. 

You can get the unit names (which are then the container names) from `fleetctl list-units`.

**Comment by anoopknayak** [comment]
Okay I tried to open up the container with `nse` command and it throws back an error
```
Error response from daemon: No such container: box_v19.ts.2.service
```
`fleetctl list-units` gives me the following:
```
UNIT					MACHINE			ACTIVE	SUB
api-beta_v2.cmd.1.service		d322a26d.../10.21.2.168	active	running
box_v19.beat.1.service			ce9f4104.../10.21.1.143	active	running
box_v19.rts.1.service			ee3daa2a.../10.21.2.231	active	running
box_v19.ss.1.service			ce9f4104.../10.21.1.143	active	running
box_v19.ss.2.service			ce9f4104.../10.21.1.143	active	running
box_v19.ts.1.service			d322a26d.../10.21.2.168	active	running
box_v19.ts.2.service			ee3daa2a.../10.21.2.231	active	running
box_v19.web.1.service			d322a26d.../10.21.2.168	active	running
clouds_v2.web.1.service			d322a26d.../10.21.2.168	active	running
deis-builder.service			ce9f4104.../10.21.1.143	active	running
deis-cache.service			ee3daa2a.../10.21.2.231	active	running
deis-controller.service			ce9f4104.../10.21.1.143	active	running
deis-database.service			d322a26d.../10.21.2.168	active	running
deis-logger.service			d322a26d.../10.21.2.168	active	running
deis-logspout.service			ce9f4104.../10.21.1.143	active	running
deis-logspout.service			d322a26d.../10.21.2.168	active	running
deis-logspout.service			ee3daa2a.../10.21.2.231	active	running
deis-publisher.service			ce9f4104.../10.21.1.143	active	running
deis-publisher.service			d322a26d.../10.21.2.168	active	running
deis-publisher.service			ee3daa2a.../10.21.2.231	active	running
deis-registry@1.service			ee3daa2a.../10.21.2.231	active	running
deis-router@1.service			d322a26d.../10.21.2.168	active	running
deis-router@2.service			ee3daa2a.../10.21.2.231	active	running
deis-router@3.service			ce9f4104.../10.21.1.143	active	running
deis-store-daemon.service		ce9f4104.../10.21.1.143	active	running
deis-store-daemon.service		d322a26d.../10.21.2.168	active	running
deis-store-daemon.service		ee3daa2a.../10.21.2.231	active	running
deis-store-gateway.service		ce9f4104.../10.21.1.143	active	running
deis-store-metadata.service		ce9f4104.../10.21.1.143	active	running
deis-store-metadata.service		d322a26d.../10.21.2.168	active	running
deis-store-metadata.service		ee3daa2a.../10.21.2.231	active	running
deis-store-monitor.service		ce9f4104.../10.21.1.143	active	running
deis-store-monitor.service		d322a26d.../10.21.2.168	active	running
deis-store-monitor.service		ee3daa2a.../10.21.2.231	active	running
deis-store-volume.service		ce9f4104.../10.21.1.143	active	running
deis-store-volume.service		d322a26d.../10.21.2.168	active	running
deis-store-volume.service		ee3daa2a.../10.21.2.231	active	running
dropbox_v2.ds.1.service			ce9f4104.../10.21.1.143	active	running
dropbox_v2.ss.1.service			ee3daa2a.../10.21.2.231	active	running
dropbox_v2.ss.2.service			d322a26d.../10.21.2.168	active	running
dropbox_v2.ts.1.service			ce9f4104.../10.21.1.143	active	running
dropbox_v2.ts.2.service			d322a26d.../10.21.2.168	active	running
dropbox_v2.web.1.service		ee3daa2a.../10.21.2.231	active	running
gdrive_v4.beat.1.service		ce9f4104.../10.21.1.143	active	running
gdrive_v4.rts.1.service			d322a26d.../10.21.2.168	active	running
gdrive_v4.ss.1.service			d322a26d.../10.21.2.168	active	running
gdrive_v4.ss.2.service			ee3daa2a.../10.21.2.231	active	running
gdrive_v4.ts.1.service			ce9f4104.../10.21.1.143	active	running
gdrive_v4.ts.2.service			d322a26d.../10.21.2.168	active	running
gdrive_v4.web.1.service			ee3daa2a.../10.21.2.231	active	running
log-forwarder.service			ce9f4104.../10.21.1.143	active	running
log-forwarder.service			d322a26d.../10.21.2.168	active	running
log-forwarder.service			ee3daa2a.../10.21.2.231	active	running
misc_v2.web.1.service			ee3daa2a.../10.21.2.231	active	running
notifications_v2.web.1.service		ce9f4104.../10.21.1.143	active	running
notifications_v2.worker.1.service	ee3daa2a.../10.21.2.231	active	running
projects_v3.services.1.service		ce9f4104.../10.21.1.143	active	running
projects_v3.services.2.service		d322a26d.../10.21.2.168	active	running
projects_v3.web.1.service		ee3daa2a.../10.21.2.231	active	running
users_v16.service.1.service		ee3daa2a.../10.21.2.231	active	running
users_v16.web.1.service			d322a26d.../10.21.2.168	active	running
users_v16.web.2.service			ce9f4104.../10.21.1.143	active	running
```

And yea is there anyway to reserve a certain amount of RAM for the container? I don't know whether it gets stuck because of memory issues. I'm downloading a file to the memory and uploading it again from the memory... Its clear that the logs get stuck and so as the process too.. `fleetctl list-units` tells me its running but I don't know what's going on in here.

**Comment by bacongobbler** [comment]
You need to be on the box which your container resides on, so you need to run `fleetctl ssh box_v19.ts.2`, then `nse box_v19.ts.2`.

> And yea is there anyway to reserve a certain amount of RAM for the container?

http://docs.deis.io/en/latest/using_deis/manage-application/#limit-the-application describes how you can limit the RAM/clock cycles for an app. More info at `deis help limits`.

**Comment by carmstrong** [comment]
@anoopknayak Did you figure this out?

**Comment by anoopknayak** [comment]
Sorry guys. I was totally busy.. I tried to get the logs in the container.. Nothing suspicious there. I tried running them on a bare docker environment in another instance. It's totally fine there. So sticking on with it. Again i had no limits set on the process. I was asking whether a chunk of memory can be reserved for a process. I still suspect the RAM might be the thing as I download data into and read it all in the memory. So closing this for now. Thanks for the help btw
